jib-core/src/main/java/com/google/cloud/tools/jib/event/EventEmitter.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2018 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.event;
+
+import com.google.common.collect.ImmutableMap;
+import java.util.Collections;
+import java.util.List;
+
+/** Emits {@link JibEvent}s to event handlers. */
+public class EventEmitter {
```

Comment:
Who creates this instance and call `.emit()`? For example, I'm roughly guessing at the beginning of `PushBlobStep.call()`, will the `call()` method directly do

`new EventEmitter(eventHandler).emit(JibEventType.BEGIN_PUSH_BLOB)`

and at the end do

`new EventEmitter(eventHandlers).emit(JibEventType.END_PUSH_BLOB)`

? If so, how could `PushBlobStep` get the `eventHandlers` from whom? Does `PushBlobStep` need to know about `eventHandlers` and pass them to `EventEmitter`? Or, will there be another manager or layer that `PushBlobStep` will use?

Reply:
The `BuildConfiguration` will have an `EventEmitter` that is created from an `EventHandlers` before execution begins. All steps will then have access to that `EventEmitter` to emit whichever events they wish.